### ISSUE TITLE
Memery leak when create  emit stream continuously

### ISSUE THREAD

**Issue by wb2110** [nkzawa/socket.io-stream#95]
I wrote a small segment code to record pcm audio  opus encode  then send to server with socket.io-stream,
After a long run ,the program run out of memory. and I detected memery leak throug memwatch, I am sure that it's this module lead this ,but I can't find the problem,anybody can help me?`
var config=require('getconfig');
var host=config.host;
var rate=config.rate;
var channels=config.channels;
var station=config.station;

var opus=require('node-opus');
var ogg=require('ogg');
var ss=require('socket.io-stream');
var miclib=require('mic');
//var Gpio=require('onoff').Gpio;
//var rec=new Gpio(92,'in','both');
var sock=require('socket.io-client');

var start=false;
var recording=false;
var sockflag=false;
var io=sock(host);
var Sock=ss(io);
var opusEncoder;
var oggEncoder;
var mic;
var sstream;
var istream;
var Random = require("random-js");
var random = new Random(Random.engines.mt19937().autoSeed());
var memwatch = require('memwatch-next');
var util=require('util');
var hd;
memwatch.on('leak', function(info) {
 console.error(info);
 if (!hd) {
   hd = new memwatch.HeapDiff();
 } else {
   var diff = hd.end();
   console.error(util.inspect(diff, true, null));
   hd = null;
 }
});
memwatch.on('stat', function(info) {
	console.log(info);
});

	
/*rec.watch(function(err,value){
	if(err){
			console.log(err);
			io.emit('error',err);
			}
			else{
				if(value==1)
				{
					console.log('begin');
					start=true
				}
				else{
					console.log('end');
					start=false;
				}
			}
	
});*/
(function randtimeout(){
	start=!start;
	setTimeout(()=>{
			randtimeout()
	},random.integer(2, 20)*100);
})();

mic=miclib({'rate':rate,'channels':channels,'debug':false});
istream=mic.getAudioStream();
	
istream.on('data',(chunk)=>{
	if(sockflag){
		if(start){
			if(!recording)
			{
				sstream=ss.createStream();
				Sock.emit('stream',sstream,{station:station});
				opusEncoder=new opus.Encoder(rate,channels);
				oggEncoder=new ogg.Encoder();
				oggEncoder.pipe(sstream);
				opusEncoder.pipe(oggEncoder.stream());
				opusEncoder.write(chunk);
				recording=true;
			}
			else{
				opusEncoder.write(chunk);
			}	
		}
		else{
			if(recording){
				opusEncoder.write(chunk);
				opusEncoder.end();
				io.emit("stop",{});
				recording=false;
			}
			else{}
			
		}
	}
});

io.on('connect',function(){
	sockflag=true;
	console.log('conected to server');	
});

Sock.on('error',(err)=>{sockflag=false;});
io.on('disconnect',function(){
	//sockflag=false;
});
mic.start();	
	var heapdump=require('heapdump');
heapdump.writeSnapshot( Date.now() + '.heapsnapshot');
setInterval(()=>{
	heapdump.writeSnapshot( Date.now() + '.heapsnapshot');
},60*60*1000);

and this is server-side code:
`   ss(client).on('stream',function(stream,message){
    	var buffers=[];
    	var start=Date.now();
    	//console.log(message);
    	stream.on('data',function(chunk){
			 buffers.push(chunk);   		
    	});
    	stream.on('end',function(){
    		//var stop=Date.now();
    		console.log('end');
    		console.log('stop',stop);
    		var buffer = Buffer.concat(buffers);
    		queue.unshift({
    			station:message.station,
    			start:start,
    			stop:stop,
    			buffer:buffer.toString('base64')
    		});
    		})
   		 });`
		

`

**Comment by SalvatoreAD** [comment]
You have to use socket.removeAllListeners () to remove all emit

**Comment by vin-ni** [comment]
So did you succeed sending audiochunks as a stream to the server?  I'm trying to do this, but can't get it to work

**Comment by wb2110** [comment]
@vin-ni yes i did, finally i use socket.io-client send the audio chunks ,it has been running for nearly a year, and runs well

**Comment by vin-ni** [comment]
Could you show me a snippet of your coder or are you just doing ‘emit‘ on the client?

And then 

    client.on('bianryData', function(data) {
      recognizeStream.write(data);
    });

**Comment by wb2110** [comment]
@vin-ni 

var miclib=require('mic');
var Gpio=require('onoff').Gpio;
var rec=new Gpio(58,'in','both');
var led1=new Gpio(config.led1,'out');
var led2=new Gpio(config.led2,'out');
var sock=require('socket.io-client');

var start=false;
var recording=false;
var sockflag=false;
var io=sock(host);
var mic;
var sstream;
var istream;
var util=require('util');
var lame=require('lame');


led2.write(1,cb);	
rec.watch(function(err,value){
	if(err){
			console.log(err);
			io.emit('error',err);
			}
			else{
				if(value==1)
				{
					console.log('begin');
					start=true
					led2.write(0,cb);
				}
				else{
					console.log('end');
					start=false;
					led2.write(1,cb);
				}
			}
	
});


mic=miclib({'rate':rate,'channels':channels,'debug':false});
istream=mic.getAudioStream();


var encoder;
istream.on('data',(chunk)=>{
	if(sockflag){
		if(start){
			if(!recording)
			{	
				var startFlag=true;
		
				var en = new lame.Encoder({
							// input 
							channels: channels,        // 2 channels (left and right) 
							bitDepth: 16,       // 16-bit samples 
							sampleRate: rate,  // 44,100 Hz sample rate 
							
							// output 
							bitRate: 128,
							outSampleRate: 22050,
							mode: lame.STEREO // STEREO (default), JOINTSTEREO, DUALCHANNEL or MONO 
						});
				encoder=en;
				en.on('data',(chunk)=>{
					if(startFlag){
						console.log('start');
						io.emit('chunk',{chunk:chunk,startFlag:true});
					}
					else{
						io.emit('chunk',{chunk:chunk});
					}
					startFlag=false;
					
				})
				en.on('end',()=>{
					startFlag=true;
					console.log('end');
					io.emit('chunk',{endFlag:true,station:station});
				});
				en.write(chunk);
				recording=true;
			}
			else{
				encoder.write(chunk);
			}	
		}
		else{
			if(recording){
				encoder.end(chunk);
				recording=false;
			}
			else{}
			
		}
	}
});

io.on('connect',function(){
	sockflag=true;
	console.log('conected to server');	
});

io.on('disconnect',()=>{
	sockflag=false;
	console.log('disconnect ');	
})


istream.on('startComplete',function(){
	console.log("fffff startComplete")
	led1.write(0,cb);
})
istream.on('processExitComplete',function(){
	console.log("fffff error")
	led1.write(1,cb);
})
function cb(){}
mic.start();

process.on('SIGINT',function(){
//rec.unexport();
led1.write(1,cb);
led2.write(1,cb);
process.exit();
})
